src/item.cpp

```diff
@@ -13222,9 +13220,22 @@ void item::set_temp_flags( units::temperature new_temperature, float freeze_perc
         set_flag( flag_COLD );
     }
 
-    // Convert water into clean water if it starts boiling
-    if( typeId() == itype_water && new_temperature > temperatures::boiling ) {
-        convert( itype_water_clean ).poison = 0;
+    // Execute temperature effects if a new threshold is reached.
+    // Water turning into clean water when boiling is a canonical example.
+    const islot_temperature_effects *te_data = type->temperature_effects_data.get();
+    if( te_data != nullptr ) {
+        for( const temperature_effect &eff : te_data->effects ) {
+            if( ( eff.temperature_above.has_value() && eff.temperature_above.value() <= new_temperature ) ||
+                ( eff.temperature_below.has_value() && eff.temperature_below.value() >= new_temperature )
+              ) {
+                if( eff.transform_into ) {
+                    convert( eff.transform_into );
+                }
+                if( eff.remove_poison ) {
```

Comment:
You probably want to do the poison *first*, in case the converted item is no longer a comestible, and thus lacks the comestible ptr. Although it would be a json error to remove_poison *and* convert it to an item without the ability to have poison, doing this in the wrong order would lead to either a crash or UB.